### PR TITLE
Fix for github issue #3790: infinite loop closing library

### PR DESCRIPTION
**Pull request by vchoi-hdfgroup**
Cause of the problem:
When h5dump tries to open the user provided test file, the metadata cache will call the "get_final_load_size" callback to find out the actual size of the the root object header.  The callback function will call H5O__prefix_deserialize() to allocate space for the object header data structure (via H5FL_CALLOC) and to deserialize the object header prefix in order to find the actual size of the object header. The metadata cache will then check whether the actual size obtained will exceed the file's EOA.
Since the actual size obtained from the test file exceeds the EOA, the metadata cache throws an error and return.
However, the oh structure that was allocated in H5O__prefix_deserialize() was not freed and hence causing the problem described in this issue. Fix:
1) Deallocate the oh structure after obtaining and saving the needed information in udata which will be used later on in the "verify_chksum" callback. 2) Deserialize the object header prefix in the "object header's "deserialize" callback regardless.  The original coding intends to keep the deserialized prefix so that the object header's "deserialize" callback does not need to deserialize the prefix again if the object header is coming through the "get_final_load_size" callback.

### CODE REVIEW

File: tools/test/h5dump/testh5dump.sh.in [hunk 1]
```diff
@@ -1561,6 +1564,9 @@ TOOLTEST2 tall-6.exp --enable-error-stack -y -o tall-6.txt -d /g1/g1.1/dset1.1.1
 # test for non-existing file
 TOOLTEST3 non_existing.ddl --enable-error-stack tgroup.h5 non_existing.h5
 
+# test to verify issue#3790: infinite loop closing library
```
Comment (derobins) [inline tools/test/h5dump/testh5dump.sh.in]: I would probably use "GitHub 3790" since 3790 could also be a Jira issue number

File: src/H5Oint.c [hunk 2]
```diff
@@ -976,7 +976,6 @@ H5O_protect(const H5O_loc_t *loc, unsigned prot_flags, bool pin_all_chunks)
     udata.v1_pfx_nmesgs           = 0;
     udata.chunk0_size             = 0;
     udata.oh                      = NULL;
-    udata.free_oh                 = false;
```
Comment (derobins) [inline src/H5Oint.c:979]: Should you initialize oh_version here? If not, it will contain garbage.

**Review by qkoziol — APPROVED** [review]
Seems like a reasonable approach to solving this specific problem.

We should consider other approaches to the general problem of "an error occurred and the cache client should clean up it's 'udata' state".  Might be good to have a new cache client callback to handle that state in the state machine.

**Comment by derobins** [comment]
Probably should have an entry in RELEASE.txt

File: tools/test/h5dump/testh5dump.sh.in (hunk 1 above)
Comment (vchoi-hdfgroup) [reply tools/test/h5dump/testh5dump.sh.in]: Fixed

File: src/H5Oint.c (hunk 2 above)
Comment (vchoi-hdfgroup) [reply src/H5Oint.c:979]: Fixed

**Comment by vchoi-hdfgroup** [comment]
> Probably should have an entry in RELEASE.txt

Fixed.

**Comment by vchoi-hdfgroup** [comment]
> Seems like a reasonable approach to solving this specific problem.
> 
> We should consider other approaches to the general problem of "an error occurred and the cache client should clean up it's 'udata' state". Might be good to have a new cache client callback to handle that state in the state machine.

Yes I agreed to consider adding a new cache client callback to handle cleaning up the 'udata' state.

**Review by bmribler — APPROVED** [review]

**Pull request merged**